### PR TITLE
app: drop stateful mempool

### PR DESCRIPTION
**Pull request by erwanor**
## Describe your changes

This is a draft PR that removes the stateful mempool from the application, replacing it with:
- a stateless mempool that gates inclusion on validity against the latest version of the app state
- a prepare proposal implementation that filters for valid transactions
- a process proposal implementation that validates block quality and lays the groundwork to do optimistic execution

## Checklist before requesting a review

- [x] If this code contains consensus-breaking changes, I have added the "consensus-breaking" label. Otherwise, I declare my belief that there are not consensus-breaking changes, for the following reason:

  > Consensus breaking

### CODE REVIEW

File: crates/core/app/src/app/mod.rs
Comment (dynst) [inline crates/core/app/src/app/mod.rs:187]:
```suggestion
        // More context in https://github.com/cometbft/cometbft/blob/v0.37.5/spec/abci/abci%2B%2B_app_requirements.md
```

The link to `abci++_app_requirements.md` added in #3643 was missing the `.md`.

Comment (erwanor) [reply crates/core/app/src/app/mod.rs:187]: Thanks!

File: crates/core/app/src/app/mod.rs
Comment (dynst) [inline crates/core/app/src/app/mod.rs:55]: What happens if a chain votes to raise block `max_bytes` to a value larger than 1MiB? This hardcoded value just doesn't allow larger blocks, right?

Comment (erwanor) [reply crates/core/app/src/app/mod.rs:55]: That's right, we don't have a flow to change consensus parameters anyway

File: crates/core/app/src/app/mod.rs
Comment (dynst) [inline crates/core/app/src/app/mod.rs:189]:
```suggestion
        // This is a node controlled parameter that is different from the homonymous
        // mempool's `max_tx_bytes`. Comet will send us raw proposals that exceed this
        // limit, presuming that a subset of those transactions will be shed.
```

'that limit' sounded like it was referring to mempool's homonymous parameter again.

File: crates/core/app/src/app/mod.rs
Comment (conorsch) [inline crates/core/app/src/app/mod.rs:61]: Thank you for taking the time to constify these values, so it's clearer that they're constant for a reason.

File: crates/core/app/src/app/mod.rs
Comment (conorsch) [inline crates/core/app/src/app/mod.rs:196]: Explicitly referring to the overloaded terminology here is greatly clarifying. :+1:

File: crates/core/app/src/server/consensus.rs
Comment (conorsch) [inline crates/core/app/src/server/consensus.rs:178]: Appreciate the forward-looking comment (https://github.com/penumbra-zone/penumbra/issues/4022).

**Review by conorsch — APPROVED** [review]

**Comment by conorsch** [comment]
Really solid work here. Especially grateful for the careful commenting throughout, as this work builds on a lot of recent and rather subtle refinements to the constraints on message sizes acceptable to the app.

**Comment by conorsch** [comment]
Not merging because review from @hdevalence is still outstanding. Once it's got final approval, @erwanor, let's use the new squash-with-commit-comments-preserved workflow we've been discussing to get it in!

File: crates/core/app/src/app/mod.rs
Comment (hdevalence) [inline crates/core/app/src/app/mod.rs:58]: Did we ever choose these in a principled way? Should they be parameters rather than hardcoded constants?

File: crates/core/app/src/server/mempool.rs
Comment (hdevalence) [inline crates/core/app/src/server/mempool.rs:18]: Why do we want the mempool to hold on to a `snapshot`? Previously, it was stateful and then it held a forked `App` state that was periodically reset, but do we need any of that now? What if the mempool just has a `Storage` handle and pulls the latest snapshot on every request?

**Review by hdevalence — APPROVED** [review]
I think we should consider making the constants be parameters, but I'm good to merge either way.

File: crates/core/app/src/app/mod.rs
Comment (erwanor) [reply crates/core/app/src/app/mod.rs:58]: This was picked with three things in mind:
- large enough to allow JIT liquidity for multiple pairs
- large enough to allow a ranged collection of positions
- large enough to allow multiple IBC client updates
- small enough to allow many of those actions to be performed in a single block even in the extreme case where each transaction is one of those things

Making them parameters make sense, I didn't do it because it would require more work with `penumbra-governance` and hooking them into https://docs.rs/tendermint/latest/tendermint/consensus/params/struct.Params.html

File: crates/core/app/src/app/mod.rs
Comment (dynst) [reply crates/core/app/src/app/mod.rs:196]: This particular helpful comment about PrepareProposal's `max_tx_bytes` was actually added in a January PR. #3643

**Pull request merged**
